phangsPipeline/casaCubeRoutines.py

```diff
@@ -26,11 +26,103 @@
+    myia=None,
+    return_myia=False,
+    return_data=False,
+    return_mask=False,
+    return_shape=False,
+    huge_cube_workaround=True,
+    ):
+    """
+    Check whether the input data cube is too large to run getchunk/putchunk globally. 
+    If so, we will need to run getchunk/putchunk channel-by-channel. 
+    
+    We will first check the cube size. If it is obviously too large, then we will directly
+    mark this case as having a memory issue. Otherwise we will try to run `getchunk` and/or 
+    `getchunk(getmask=True)` to see whether it returns a cube data array. In the case of 
+    having a memory issue, `getchunk` will return a flattend array whose shape is different
+    from the input data cube.
+    
+    This memory issue is also noted in https://casa.nrao.edu/docs/casaref/image.putchunk.html 
+    as: "If all the pixels didn't easily fit in memory, you would iterate through the image 
+    chunk by chunk to avoid exhausting virtual memory."
+    
+    Args:
+        infile: input image file.
+        myia: optional, if not given then we will run `myia = au.createCasaTool(casaStuff.iatool)` then `myia.open(infile)`.
+        return_data: if True then we will return the data if `getchunk()` runs properly withut a memory issue.
+        return_mask: if True then we will return the mask if `getchunk(getmask=True)` runs properly withut a memory issue.
+        return_shape: if True then we will return the shape of the cube data.
+        huge_cube_workaround: historical arg for compatibility. Setting it to False will disable per-channel processing and may cause error.
+    
+    Returns:
+        Tuple of 1-5 variables: 
+            check_no_memory_issue (boolean), 
+            myia object (if return_myia is True),
+            data array (if return_data is True),
+            mask array (if return_mask is True),
+            shape list (if return_shape is True)
+    """
+    has_memory_issue = False
+    cube_data = None
+    cube_mask = None
+    has_opened_file = False
+    if myia is None:
+        myia = au.createCasaTool(casaStuff.iatool)
+        myia.open(infile)
+        has_opened_file = True
+    if not myia.isopen():
+        myia.open(infile)
+        has_opened_file = True
+    cube_shape = myia.shape() # [X, Y, CHANNEL, [STOKES]]
+    if np.prod(cube_shape) >= 2880*2880*393: # known memory issue
+        has_memory_issue = True
+    if not has_memory_issue: # try to see if there is a memory issue
+        # in which case the getchunk will return a flat array instead of the cube shape
+        if return_data or not return_mask:
+            cube_data = myia.getchunk() # data.shape = [X, Y, CHANNEL, [STOKES]]
+            check_shape = cube_data.shape
+        if return_mask:
+            cube_mask = myia.getchunk(getmask=True) # mask.shape = [X, Y, CHANNEL, [STOKES]]
+            check_shape = cube_mask.shape
+        if not np.all(check_shape == cube_shape): # shape does not match, has memory issue
```

Comment:
I didn't fully follow why this function can also return the whole cube or mask. Is this the key line? That we need to be able to fit the mask or data into memory to have the correct shape?

Reply:
This is mainly for the efficiency. We will anyway run a `getchunk` to test if it returns a correct array size or not. I let the code to run `getchunk()` with getmask=True if return_mask, or getmask=False if return_data, so that if it is executed successfully, we can immediately return the result as the user wanted. If not doing so, we will need to rerun getchunk in the later part of the code to get what we want in `casaCubeRoutines.get_mask`, `casaCubeRoutines.copy_mask`, `casaCubeRoutines.multiply_cube_by_value`, and `casaFeatherRoutines.feather_two_cubes`, and waste several to tens of minutes. 

In the case that there is a memory issue for big data array, the code will set has_memory_issue=True, and will set the returned data or mask to None even if return_data or return_mask has been set. The later part of the code will check has_memory_issue and do per-channel computations correspondingly.